pkg/agent/cniserver/interface_configuration_linux.go

```diff
@@ -91,49 +92,154 @@ func (ic *ifConfigurator) moveIfToNetns(ifname string, netns ns.NetNS) error {
+	if err != nil {
+		return fmt.Errorf("failed to open container netns %s: %w", containerNetNS, err)
+	}
+	defer netns.Close()
+
+	if err = tempNS.Do(func(hostNS ns.NetNS) error {
+		// Look up the device in tempNS, as the index might have changed
+		tempNSDev, err := ic.netlink.LinkByName(vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set MTU for VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to find %s in tempNS: %w", vfNetDevice, err)
 		}
-		err = ic.netlink.LinkSetUp(link)
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetNsFd(tempNSDev, int(hostNS.Fd()))
+			}
+		}()
+
+		// Rename the device to the wanted name
+		if err = ic.netlink.LinkSetName(tempNSDev, containerIfaceName); err != nil {
+			return fmt.Errorf("failed to rename VF device %s to %s: %w", vfNetDevice, containerIfaceName, err)
+		}
+
+		// Restore the original device name in case of error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetName(tempNSDev, vfNetDevice)
+			}
+		}()
+
+		err = ic.netlink.LinkSetAlias(tempNSDev, vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set link up to VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to set alias as %s for VF netdevice %s: %w", vfNetDevice, vfNetDevice, err)
 		}
-		containerIface.Name = containerIfaceName
-		containerIface.Mac = link.Attrs().HardwareAddr.String()
-		containerIface.Sandbox = netns.Path()
-		klog.V(2).Infof("hostIface: %+v, containerIface: %+v", hostIface, containerIface)
-		klog.V(2).Infof("Configuring IP address for container %s", containerID)
-		// result.Interfaces must be set before this.
-		if err := ipamConfigureIface(containerIface.Name, result); err != nil {
-			return fmt.Errorf("failed to configure IP address for container %s: %v", containerID, err)
+		klog.V(2).InfoS("Link's alias has been updated", "alias", vfNetDevice)
+
+		// Remove the alias on error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				_ = ic.netlink.LinkSetAlias(tempNSDev, "")
+			}
+		}()
+
+		// Move VF to Container namespace
+		err = ic.moveIFToNetNS(containerIfaceName, netns)
+		if err != nil {
+			return fmt.Errorf("failed to move VF %s to container netns %s: %w", vfNetDevice, containerNetNS, err)
 		}
-		klog.V(2).Infof("ipam.ConfigureIface result: %+v, err: %v", result, err)
-		return nil
+
+		// Look up the device again on error, as the index might have changed
+		defer func() {
+			if err != nil {
+				tempNSDev, _ = ic.netlink.LinkByName(containerIfaceName)
+				klog.InfoS("Look up the device again on error, as the index might have changed")
+			}
+		}()
+
+		err = netns.Do(func(_ ns.NetNS) error {
+			link, err := ic.netlink.LinkByName(containerIfaceName)
+			if err != nil {
+				return fmt.Errorf("failed to find VF netdevice %s: %v", containerIfaceName, err)
+			}
+			// Move the interface back to tempNS on error
```

Comment:
Is this step necessary? Could we use variable to save the current interface netns instead?

Reply:
We need it to tempNS to undo the operation like resetting the interface name and alias if there is an error.

Reply:
Ok, could we add info the to comment?